File: android/app/src/main/java/com/laneshadow/ui/molecules/LSFormField.kt
package com.laneshadow.ui.molecules

import androidx.compose.foundation.layout.Column
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.ui.text.input.VisualTransformation
import com.laneshadow.theme.generated.LaneShadowTheme.IconName
import com.laneshadow.theme.LocalLaneShadowTheme
import com.laneshadow.ui.atoms.InputState
import com.laneshadow.ui.atoms.LSText
import com.laneshadow.ui.atoms.LSTextField
import com.laneshadow.ui.atoms.ContentColor
import com.laneshadow.ui.atoms.TypographyVariant

/**
 * LSFormField molecule component
 *
 * Form field molecule composing LSText label, LSTextField atom, and optional error text.
 * Follows the design spec at .spec/design/system/molecules/form-field/
 *
 * @param label Field label text
 * @param value Current input value
 * @param onValueChange Callback when input value changes
 * @param error Optional error message (non-null shows error state)
 * @param placeholder Optional placeholder text
 * @param modifier Modifier for the form field container
 */
@Composable
fun LSFormField(
    label: String,
    value: String,
    onValueChange: (String) -> Unit,
    error: String? = null,
    helper: String? = null,
    placeholder: String? = null,
    state: InputState = InputState.Default,
    leadingIcon: IconName? = null,
    trailingIcon: IconName? = null,
    enabled: Boolean = true,
    keyboardOptions: KeyboardOptions = KeyboardOptions.Default,
    keyboardActions: KeyboardActions = KeyboardActions.Default,
    visualTransformation: VisualTransformation = VisualTransformation.None,
    inputModifier: Modifier = Modifier,
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current

    Column(
        modifier = modifier,
        verticalArrangement = androidx.compose.foundation.layout.Arrangement.spacedBy(theme.space.sm),
    ) {
        // Label
        LSText(
            text = label,
            variant = TypographyVariant.Ui.Label.Md,
            color = ContentColor.Primary,
        )

        // Input field with error state
        LSTextField(
            value = value,
            onValueChange = onValueChange,
            state = when {
                !enabled -> InputState.Disabled
                error != null -> InputState.Error
                else -> state
            },
            placeholder = placeholder,
            leadingIcon = leadingIcon,
            trailingIcon = trailingIcon,
            modifier = inputModifier,
            keyboardOptions = keyboardOptions,
            keyboardActions = keyboardActions,
            visualTransformation = visualTransformation,
        )

        // Error text (only when error is non-null)
        if (error != null) {
            LSText(
                text = error,
                variant = TypographyVariant.Ui.Body.Sm,
                color = ContentColor.Error,
            )
        } else if (helper != null) {
            LSText(
                text = helper,
                variant = TypographyVariant.Ui.Body.Sm,
                color = ContentColor.Subtle,
            )
        }
    }
}

File: ios/LaneShadow/Views/Molecules/LSFormField.swift
import LaneShadowTheme
import SwiftUI

public struct LSFormField: View {
    @Environment(\.theme) private var theme

    @Binding private var value: String
    private let label: String
    private let placeholder: String?
    private let error: String?
    private let helperText: String?
    private let isRequired: Bool
    private let state: InputState
    private let isSecureEntry: Bool
    private let leadingIcon: IconName?
    private let trailingIcon: IconName?
    private let leadingSymbolName: String?
    private let trailingSymbolName: String?
    private let inputAccessibilityIdentifier: String?

    public init(
        label: String,
        value: Binding<String>,
        placeholder: String? = nil,
        error: String? = nil,
        helperText: String? = nil,
        isRequired: Bool = false,
        state: InputState = .default,
        isSecureEntry: Bool = false,
        leadingIcon: IconName? = nil,
        trailingIcon: IconName? = nil,
        leadingSymbolName: String? = nil,
        trailingSymbolName: String? = nil,
        inputAccessibilityIdentifier: String? = nil
    ) {
        self.label = label
        _value = value
        self.placeholder = placeholder
        self.error = error
        self.helperText = helperText
        self.isRequired = isRequired
        self.state = state
        self.isSecureEntry = isSecureEntry
        self.leadingIcon = leadingIcon
        self.trailingIcon = trailingIcon
        self.leadingSymbolName = leadingSymbolName
        self.trailingSymbolName = trailingSymbolName
        self.inputAccessibilityIdentifier = inputAccessibilityIdentifier
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: theme.space.sm) {
            // Label row
            HStack(spacing: theme.space.xs) {
                LSText(label, variant: .label.md, color: .primary)

                if isRequired {
                    LSText("*", variant: .label.md, color: .danger)
                }
            }

            // Input field
            LSTextField(
                value: $value,
                placeholder: placeholder,
                state: error != nil ? .error : state,
                isSecureEntry: isSecureEntry,
                leadingIcon: leadingIcon,
                trailingIcon: trailingIcon,
                leadingSymbolName: leadingSymbolName,
                trailingSymbolName: trailingSymbolName,
                helperText: error == nil ? helperText : nil,
                inputAccessibilityIdentifier: inputAccessibilityIdentifier
            )

            // Error text
            if let error {
                HStack(spacing: theme.space.xs) {
                    LSIcon(name: .close, size: .xs, resolvedColorOverride: theme.colors.danger.default)
                    LSText(error, variant: .body.sm, color: .danger)
                }
            }
        }
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("lsformfield")
    }
}
